File: SignalAndSystem/exp2.py
import matplotlib.pyplot as plt
import numpy as np


def q1():
    E = 1.2
    T = 0.1  # ms
    tau = 1 / 4 * T

    num_samples = 256
    t = np.linspace(0, T, num_samples)

    f_t = np.zeros_like(t)
    f_t[t < tau] = E

    harmonics = 21  # 0 to 20
    frequencies = np.arange(harmonics) / T

    a0 = E * tau / T  # DC
    amplitudes = np.zeros(harmonics)
    amplitudes[0] = a0

    for n in range(1, harmonics):
        # a_n = E * tau / T * sin(n*pi*tau/T) / (n*pi*tau/T)
        arg = n * np.pi * tau / T
        amplitudes[n] = E * tau / T * np.sin(arg) / arg * 2  # 乘以2是因为有正负两边

    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 8))

    ax1.plot(t, f_t, "b-", linewidth=2)
    ax1.set_xlim(0, T)
    ax1.set_ylim(-0.2, 1.5)
    ax1.set_xlabel("t (s)")
    ax1.set_ylabel("f(t) (V)")
    ax1.set_title("Exp2Q1 - Signal")
    ax1.grid(True, alpha=0.3)

    ax2.stem(frequencies, amplitudes, basefmt=" ")
    ax2.set_xlim(0, 20 / T)
    ax2.set_ylim(-0.25, 0.75)
    ax2.set_xlabel("Frequency (Hz)")
    ax2.set_ylabel("F(t)")
    ax2.set_title("Freq. Spec.")
    ax2.grid(True, alpha=0.3)

    plt.tight_layout()
    plt.savefig("output/exp2q1.png")
    plt.close()


def q2():
    t = np.linspace(0, 10, 100)
    f_t = np.zeros(100)

    # delta(t-3):
    f_t[30] = 1
    freq = np.linspace(0, 5, 100)
    magnitude = np.ones_like(freq)

    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 8))

    ax1.plot(t, f_t, "b-", linewidth=2)
    ax1.set_xlim(0, 10)
    ax1.set_ylim(-0.1, 1.1)
    ax1.set_xlabel("t (s)")
    ax1.set_ylabel("f(t)")
    ax1.set_title("Exp2Q2 - Signal")
    ax1.grid(True, alpha=0.3)

    ax2.plot(freq, magnitude, "r-")
    ax2.set_xlim(0, 5)
    ax2.set_ylim(-1, 2)
    ax2.set_xlabel("Frequency (Hz)")
    ax2.set_ylabel("F(t)")
    ax2.set_title("Freq. Spec.")
    ax2.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig("output/exp2q2.png")
    plt.close()


def q3():
    f = 1  # Hz
    T = 1 / f

    num_samples = 200
    t = np.linspace(0, T, num_samples, endpoint=False)

    # Triangular
    period = t % T
    f_t = np.where(period < T / 2, 16 * period - 4, -16 * (period - T / 2) + 4)

    harmonics = np.arange(-20, 21)  # -20 to 20
    amplitudes = np.zeros(len(harmonics))
    for k in range(21):
        n = 2 * k + 1
        if n <= 20:
            sign = (-1) ** k
            amp = 32 / (np.pi**2 * n**2) * sign
            idx_neg = 20 - n
            idx_pos = 20 + n
            amplitudes[idx_neg] = amp
            amplitudes[idx_pos] = amp

    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 8))

    ax1.plot(t, f_t, "b-", linewidth=2)
    ax1.set_xlim(0, T)
    ax1.set_ylim(-5, 5)
    ax1.set_xlabel("t (s)")
    ax1.set_ylabel("f(t)")
    ax1.set_title("Exp2Q3 - Signal")
    ax1.grid(True, alpha=0.3)

    ax2.stem(harmonics, amplitudes, basefmt=" ")
    ax2.set_xlim(-20, 20)
    ax2.set_ylim(-0.5, 2)
    ax2.set_xlabel("Harmonic")
    ax2.set_ylabel("F(t)")
    ax2.set_title("Freq. Spec.")
    ax2.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig("output/exp2q3.png")
    plt.close()


def q4():
    E = 1
    f = 5000  # Hz
    T = 1 / f
    num_harmonics = 20

    t = np.linspace(0, 2 * T, 200)
    harmonics = np.arange(1, num_harmonics + 1)

    T_grid, H_grid = np.meshgrid(t, harmonics)

    # 计算每个谐波在每个时刻的值
    Z = np.zeros_like(T_grid, dtype=float)
    dc_component = E / 2

    for i, n in enumerate(harmonics):
        Z[i, :] = dc_component - (E / (n * np.pi)) * np.sin(2 * np.pi * n * f * t)

    # 计算谐波叠加（逐步累加）
    accumulated = np.zeros_like(T_grid, dtype=float)
    accumulated[0, :] = Z[0, :]
    for i in range(1, num_harmonics):
        accumulated[i, :] = accumulated[i - 1, :] + Z[i, :]

    # 3D网格图
    fig = plt.figure(figsize=(16, 9))
    ax = fig.add_subplot(111, projection="3d")

    surf = ax.plot_surface(T_grid * 1e6, H_grid, accumulated, cmap="viridis", alpha=0.9)

    ax.set_xlabel("Time (μs)")
    ax.set_ylabel("Harmonic Order")
    ax.set_zlabel("f(t)")
    ax.set_title("Exp2Q4")

    fig.colorbar(surf, ax=ax, label="Amplitude")
    plt.savefig("output/exp2q4.png", bbox_inches="tight")
    plt.close()


def q5():
    E = 1
    f1 = 1000
    T1 = 1 / f1
    tho = 1 / 4 * T1
    num_harmonics = 20

    t = np.linspace(0, 2 * T1, 200)
    harmonics = np.arange(1, num_harmonics + 1)

    T_grid, H_grid = np.meshgrid(t, harmonics)

    # 计算每个谐波在每个时刻的值
    Z = np.zeros_like(T_grid, dtype=float)

    for i, n in enumerate(harmonics):
        # 矩形波的谐波：a_n cos(2pi n f1 t)
        # a_n = (2 * E * tho / T1) * sinc(n * tho / T1)
        # sinc(x) = sin(pi x)/(pi x), 但这里 sin(n pi tho / T1) / (n pi tho / T1) = sinc(n tho / T1)
        a_n = (2 * E * tho / T1) * np.sinc(n * tho / T1)
        Z[i, :] = a_n * np.cos(2 * np.pi * n * f1 * t)

    # 计算谐波叠加（逐步累加）
    accumulated = np.zeros_like(T_grid, dtype=float)
    accumulated[0, :] = Z[0, :]
    for i in range(1, num_harmonics):
        accumulated[i, :] = accumulated[i - 1, :] + Z[i, :]

    # 3D网格图
    fig = plt.figure(figsize=(16, 9))
    ax = fig.add_subplot(111, projection="3d")

    surf = ax.plot_surface(T_grid * 1e6, H_grid, accumulated, cmap="viridis", alpha=0.9)

    ax.set_xlabel("Time (μs)")
    ax.set_ylabel("Harmonic Order")
    ax.set_zlabel("f(t)")
    ax.set_title("Exp2Q5")

    fig.colorbar(surf, ax=ax, label="Amplitude")
    plt.savefig("output/exp2q5.png", dpi=100, bbox_inches="tight")
    plt.close()


if __name__ == "__main__":
    q1()
    q2()
    q3()
    q4()
    q5()

File: SignalAndSystem/exp2.m
function exp2()
    % Exp2: 连续时间信号的傅里叶分析
    if ~exist('output', 'dir')
        mkdir('output');
    end

    q1();
    q2();
    q3();
    q4();
    q5();
end

function q1()
    E = 1.2;
    T = 0.1; % ms? code says T=0.1
    tau = 1/4 * T;

    num_samples = 256;
    t = linspace(0, T, num_samples);

    f_t = zeros(size(t));
    f_t(t < tau) = E;

    harmonics = 21; % 0 to 20
    frequencies = (0:harmonics-1) / T;

    a0 = E * tau / T; % DC
    amplitudes = zeros(1, harmonics);
    amplitudes(1) = a0;

    for n = 1:(harmonics-1)
        % n here corresponds to index n+1 in MATLAB (since 1 is DC)
        % BUT loop variable n is harmonic number
        arg = n * pi * tau / T;
        val = E * tau / T * sin(arg) / arg * 2;
        amplitudes(n+1) = val;
    end

    f = figure('Visible', 'off', 'Position', [0 0 1200 800]);
    subplot(2, 1, 1);
    plot(t, f_t, 'b-', 'LineWidth', 2);
    xlim([0, T]);
    ylim([-0.2, 1.5]);
    xlabel('t (s)');
    ylabel('f(t) (V)');
    title('Exp2Q1 - Signal');
    grid on;

    subplot(2, 1, 2);
    stem(frequencies, amplitudes, 'BaseValue', 0);
    xlim([0, 20/T]);
    ylim([-0.25, 0.75]);
    xlabel('Frequency (Hz)');
    ylabel('F(t)');
    title('Freq. Spec.');
    grid on;

    saveas(f, 'output/exp2q1.png');
    close(f);
end

function q2()
    t = linspace(0, 10, 100);
    f_t = zeros(1, 100);

    % delta(t-3):
    % Python: f_t[30] = 1. In MATLAB index 31 corresponds to 30 if 0-based.
    % linspace 0 to 10 with 100 points.
    % indices: 1..100
    % t(31) is close to 3?
    % t = 0, 0.101..., 10.
    % Let's just set the 31st element to 1 to match python approximately
    f_t(31) = 1;

    freq = linspace(0, 5, 100);
    magnitude = ones(size(freq));

    f = figure('Visible', 'off', 'Position', [0 0 1200 800]);

    subplot(2, 1, 1);
    plot(t, f_t, 'b-', 'LineWidth', 2);
    xlim([0, 10]);
    ylim([-0.1, 1.1]);
    xlabel('t (s)');
    ylabel('f(t)');
    title('Exp2Q2 - Signal');
    grid on;

    subplot(2, 1, 2);
    plot(freq, magnitude, 'r-');
    xlim([0, 5]);
    ylim([-1, 2]);
    xlabel('Frequency (Hz)');
    ylabel('F(t)');
    title('Freq. Spec.');
    grid on;

    saveas(f, 'output/exp2q2.png');
    close(f);
end

function q3()
    frequency = 1; % Hz
    T = 1 / frequency;

    num_samples = 200;
    % linspace(0, T, num_samples) includes endpoint usually.
    % Python: np.linspace(0, T, num_samples, endpoint=False)
    % MATLAB: linspace includes endpoint unless we manipulate.
    t = linspace(0, T, num_samples+1);
    t(end) = []; % Remove last point to act like endpoint=False

    % Triangular
    % Python: period = t % T
    % f_t = np.where(period < T / 2, 16 * period - 4, -16 * (period - T / 2) + 4)
    period_t = mod(t, T);
    f_t = zeros(size(t));
    mask = period_t < T/2;
    f_t(mask) = 16 * period_t(mask) - 4;
    f_t(~mask) = -16 * (period_t(~mask) - T/2) + 4;

    harmonics = -20:20;
    amplitudes = zeros(size(harmonics));

    % Python: loop k in range(21), n = 2*k+1.
    for k = 0:20
        n = 2 * k + 1;
        if n <= 20
            sign_val = (-1)^k;
            amp = 32 / (pi^2 * n^2) * sign_val;

            % Python: idx_neg = 20 - n; idx_pos = 20 + n
            % MATLAB indices for harmonics array:
            % harmonics array is -20, -19, ..., 0, ..., 20. (Size 41)
            % 0 corresponds to index 21. n corresponds to index 21+n. -n to 21-n.
            idx_neg = 21 - n;
            idx_pos = 21 + n;

            amplitudes(idx_neg) = amp;
            amplitudes(idx_pos) = amp;
        end
    end

    f = figure('Visible', 'off', 'Position', [0 0 1200 800]);

    subplot(2, 1, 1);
    plot(t, f_t, 'b-', 'LineWidth', 2);
    xlim([0, T]);
    ylim([-5, 5]);
    xlabel('t (s)');
    ylabel('f(t)');
    title('Exp2Q3 - Signal');
    grid on;

    subplot(2, 1, 2);
    stem(harmonics, amplitudes, 'BaseValue', 0);
    xlim([-20, 20]);
    ylim([-0.5, 2]);
    xlabel('Harmonic');
    ylabel('F(t)');
    title('Freq. Spec.');
    grid on;

    saveas(f, 'output/exp2q3.png');
    close(f);
end

function q4()
    E = 1;
    freq = 5000; % Hz
    T = 1 / freq;
    num_harmonics = 20;

    t = linspace(0, 2 * T, 200);
    harmonics_vec = 1:num_harmonics;

    [T_grid, H_grid] = meshgrid(t, harmonics_vec);

    Z = zeros(size(T_grid));
    dc_component = E / 2;

    % harmonics_vec is 1..20
    % H_grid rows are constant harmonic number

    for i = 1:num_harmonics
        n = harmonics_vec(i);
        Z(i, :) = dc_component - (E / (n * pi)) * sin(2 * pi * n * freq * t);
    end

    accumulated = zeros(size(T_grid));
    accumulated(1, :) = Z(1, :);
    for i = 2:num_harmonics
        accumulated(i, :) = accumulated(i - 1, :) + Z(i, :);
    end

    f = figure('Visible', 'off', 'Position', [0 0 1600 900]);
    ax = axes('Parent', f);
    surf(ax, T_grid * 1e6, H_grid, accumulated, 'EdgeColor', 'none', 'FaceAlpha', 0.9);
    colormap(ax, 'parula'); % 'viridis' not standard in old MATLAB, parula is good
    colorbar(ax);

    xlabel(ax, 'Time (\mus)');
    ylabel(ax, 'Harmonic Order');
    zlabel(ax, 'f(t)');
    title(ax, 'Exp2Q4');
    view(ax, 3);

    saveas(f, 'output/exp2q4.png');
    close(f);
end

function q5()
    E = 1;
    f1 = 1000;
    T1 = 1 / f1;
    tho = 1/4 * T1;
    num_harmonics = 20;

    t = linspace(0, 2 * T1, 200);
    harmonics_vec = 1:num_harmonics;

    [T_grid, H_grid] = meshgrid(t, harmonics_vec);

    Z = zeros(size(T_grid));

    for i = 1:num_harmonics
        n = harmonics_vec(i);
        % sinc(x) = sin(pi*x)/(pi*x).
        % Python: np.sinc(n * tho / T1) => sin(pi * n * tho / T1) / (pi * n * tho / T1)
        % This matches MATLAB sinc definition.

        a_n = (2 * E * tho / T1) * sinc(n * tho / T1);
        Z(i, :) = a_n * cos(2 * pi * n * f1 * t);
    end

    accumulated = zeros(size(T_grid));
    accumulated(1, :) = Z(1, :);
    for i = 2:num_harmonics
        accumulated(i, :) = accumulated(i - 1, :) + Z(i, :);
    end

    f = figure('Visible', 'off', 'Position', [0 0 1600 900]);
    ax = axes('Parent', f);
    surf(ax, T_grid * 1e6, H_grid, accumulated, 'EdgeColor', 'none', 'FaceAlpha', 0.9);
    colormap(ax, 'parula');
    colorbar(ax);

    xlabel(ax, 'Time (\mus)');
    ylabel(ax, 'Harmonic Order');
    zlabel(ax, 'f(t)');
    title(ax, 'Exp2Q5');
    view(ax, 3);

    saveas(f, 'output/exp2q5.png');
    close(f);
end
